### PR TITLE
[SPARK-40313][PS] Make `ps.DataFrame(data, index)` support the same anchor

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Make `ps.DataFrame(data, index)` support the same anchor

### Why are the changes needed?
before:
```
In [1]: import pyspark.pandas as ps

In [2]: psdf = ps.DataFrame([[1,2], [3,4]], columns=["A", "B"])

In [3]: ps.DataFrame(data=psdf[["A"]] * 2, index=psdf.index)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
...
ValueError: Cannot combine the series or dataframe because it comes from a different dataframe. In order to allow this operation, enable 'compute.ops_on_diff_frames' option.
```

after:
```
In [1]: import pyspark.pandas as ps

In [2]: psdf = ps.DataFrame([[1,2], [3,4]], columns=["A", "B"])

In [3]: ps.DataFrame(data=psdf[["A"]] * 2, index=psdf.index)
                                                                                
   A
0  2
1  6

```

### Does this PR introduce _any_ user-facing change?
yes, DataFrame creation now support same anchor

### How was this patch tested?
added UT

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]
Looks fine. Cc @ueshin

File: python/pyspark/pandas/tests/test_dataframe.py [hunk 1]
```diff
@@ -343,6 +343,80 @@ def test_creation_index(self):
...
+            pd.DataFrame(data=pdf, index=pdf.index),
+        )
+        self.assert_eq(
+            ps.DataFrame(data=psdf + 1, index=psdf.index),
+            pd.DataFrame(data=pdf + 1, index=pdf.index),
+        )
+        self.assert_eq(
+            ps.DataFrame(data=psdf[["y"]] * 2, index=psdf.index),
+            pd.DataFrame(data=pdf[["y"]] * 2, index=pdf.index),
+        )
```
Comment (itholic) [inline python/pyspark/pandas/tests/test_dataframe.py:418]: Can we test it with the `MultiIndex` and other index types such as `CategoricalIndex` and `TimedeltaIndex` ?

**Review by itholic — COMMENTED** [review]
Looks good otherwise!

File: python/pyspark/pandas/tests/test_dataframe.py (hunk 1 above)
Comment (zhengruifeng) [reply python/pyspark/pandas/tests/test_dataframe.py:418]: let me add them

Comment (zhengruifeng) [reply python/pyspark/pandas/tests/test_dataframe.py:418]: Done

Comment (itholic) [reply python/pyspark/pandas/tests/test_dataframe.py:418]: Cool!

**Review by itholic — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by zhengruifeng** [comment]
Thank you @itholic @HyukjinKwon for reivews

**Pull request closed without merging**